cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DefaultProgramRunnerFactory.java

```diff
@@ -40,30 +50,81 @@
   private final Map<ProgramType, Provider<ProgramRunner>> defaultRunnerProviders;
   private final ProgramRuntimeProvider.Mode mode;
   private final ProgramRuntimeProviderLoader runtimeProviderLoader;
-
+  private final ProgramStateWriter programStateWriter;
 
   @Inject
   DefaultProgramRunnerFactory(Injector injector, ProgramRuntimeProvider.Mode mode,
                               ProgramRuntimeProviderLoader runtimeProviderLoader,
-                              Map<ProgramType, Provider<ProgramRunner>> defaultRunnerProviders) {
+                              Map<ProgramType, Provider<ProgramRunner>> defaultRunnerProviders,
+                              ProgramStateWriter programStateWriter) {
     this.injector = injector;
     this.defaultRunnerProviders = defaultRunnerProviders;
     this.mode = mode;
     this.runtimeProviderLoader = runtimeProviderLoader;
+    this.programStateWriter = programStateWriter;
   }
 
   @Override
   public ProgramRunner create(ProgramType programType) {
     ProgramRuntimeProvider provider = runtimeProviderLoader.get(programType);
+    ProgramRunner runner;
+
     if (provider != null) {
       LOG.debug("Using runtime provider {} for program type {}", provider, programType);
-      return provider.createProgramRunner(programType, mode, injector);
+      runner = provider.createProgramRunner(programType, mode, injector);
+    } else {
+      Provider<ProgramRunner> defaultProvider = defaultRunnerProviders.get(programType);
+      if (defaultProvider == null) {
+        throw new IllegalArgumentException("Unsupported program type: " + programType);
+      }
+      runner = defaultProvider.get();
+    }
+
+    // Wrap a state change listener around the controller returned by the program runner in standalone mode
+    // In distributed mode, program state changes are recorded in the event handler by Twill AM.
+    return (mode == ProgramRuntimeProvider.Mode.LOCAL)
+      ? new LocalProgramRunner(runner, programStateWriter)
```

Comment:
SparkTwillRunnable doesn't use the program runner factory, so the SparkTwillRunnable produces a SparkProgramRunner - this doesn't persist the state change. This is good because the Spark programs in distributed mode persist state changes from Twill AM.

WorkflowTwillRunnable does bind to local mode, so it's Spark or MR programs will have the inner program runner that does persist state change.